asg/src/reducer/monoid/tuple.rs

```diff
@@ -0,0 +1,161 @@
+// Copyright (C) 2019-2021 Aleo Systems Inc.
+// This file is part of the Leo library.
+
+// The Leo library is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// The Leo library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with the Leo library. If not, see <https://www.gnu.org/licenses/>.
+use super::*;
+
+macro_rules! tuple_append {
```

Comment:
This macro seems like it could be cleaned up with a head and tails pattern matching.

Reply:
I may need some help with this. 
I took a stab at generating tuple indices in declarative macros but had some difficulty getting it to work
Generally this seems to be a hard thing to [do].(https://stackoverflow.com/questions/66396814/generating-tuple-indices-based-on-macro-rules-repetition-expansion)

Reply:
Sure I can help take a look at some point